src/js/actions/tools.js

```diff
@@ -600,7 +602,29 @@ define(function (require, exports) {
                 .then(function (policyID) {
                     this.dispatch(events.tool.VECTOR_MASK_POLICY_CHANGE, policyID);
                 });
+            removePromise = Promise.resolve();
         } else {
+            // delete empty masks when leaving mask mode
+            removePromise = this.transfer(layers.resetLayers, currentDocument, currentLayer)
+                .then(function () {
+                    currentLayer = appStore.getCurrentDocument().layers.selected.first();
+
+                    if (currentLayer.vectorMaskEnabled && currentLayer.vectorMaskEmpty) {
+                        return descriptor.playObject(vectorMaskLib.deleteVectorMask());
+                    } else {
+                        return Promise.resolve();
+                    }
+                })
+                .bind(this)
+                .then(function () {
+                    var payload = {
+                            documentID: currentDocument.id,
+                            layerIDs: Immutable.List.of(currentLayer.id),
+                            vectorMaskEnabled: false
+                        };
+                    return this.dispatch(events.document.history.optimistic.REMOVE_VECTOR_MASK_FROM_LAYER, payload);
```

Comment:
It seems like this will dispatch *always* when exiting vector mode, which is not desirable for existing+unchanged vector modes.